From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Wed, 17 Dec 2025 10:28:36 -0800
Subject: [PATCH] Purpur Ender Chest 6 Rows Config


diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 5fda54577b7c54f793f072b4d9b09747d8448d66..92cd4c3c619b2ecd719a01edcf5b2b2347fbbcfa 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -947,6 +947,27 @@ public abstract class PlayerList {
         player.getBukkitEntity().recalculatePermissions(); // CraftBukkit
         this.server.getCommands().sendCommands(player);
         } // Paper - Add sendOpLevel API
+        // Purpur start - Barrels and enderchests 6 rows
+
+        if (io.canvasmc.canvas.Config.INSTANCE.containers.enderChestSixRows && io.canvasmc.canvas.Config.INSTANCE.containers.enderChestPermissionRows) {
+            org.bukkit.craftbukkit.entity.CraftHumanEntity bukkit = player.getBukkitEntity();
+            if (bukkit.hasPermission("purpur.enderchest.rows.six")) {
+                player.sixRowEnderchestSlotCount = 54;
+            } else if (bukkit.hasPermission("purpur.enderchest.rows.five")) {
+                player.sixRowEnderchestSlotCount = 45;
+            } else if (bukkit.hasPermission("purpur.enderchest.rows.four")) {
+                player.sixRowEnderchestSlotCount = 36;
+            } else if (bukkit.hasPermission("purpur.enderchest.rows.three")) {
+                player.sixRowEnderchestSlotCount = 27;
+            } else if (bukkit.hasPermission("purpur.enderchest.rows.two")) {
+                player.sixRowEnderchestSlotCount = 18;
+            } else if (bukkit.hasPermission("purpur.enderchest.rows.one")) {
+                player.sixRowEnderchestSlotCount = 9;
+            }
+        } else {
+            player.sixRowEnderchestSlotCount = -1;
+        }
+        // Purpur end - Barrels and enderchests 6 rows
     }
 
     // Paper start - whitelist verify event / login event
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index bf9b97788909875bf6b12f24987783e18f0af2e1..0b92a9e897336baa7c17976788c08555ecc34997 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -179,6 +179,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     private int currentImpulseContextResetGraceTime = 0;
     public boolean affectsSpawning = true; // Paper - Affects Spawning API
     public net.kyori.adventure.util.TriState flyingFallDamage = net.kyori.adventure.util.TriState.NOT_SET; // Paper - flying fall damage
+	 public int sixRowEnderchestSlotCount = -1; // Purpur - Barrels and enderchests 6 rows
 
     // CraftBukkit start
     public boolean fauxSleeping;
diff --git a/net/minecraft/world/inventory/ChestMenu.java b/net/minecraft/world/inventory/ChestMenu.java
index 72bde309c917448edc8f12b43348c4dcf7c8b117..68b406cdcd1571314d2f0ee53b0e78ddd1c4664e 100644
--- a/net/minecraft/world/inventory/ChestMenu.java
+++ b/net/minecraft/world/inventory/ChestMenu.java
@@ -66,10 +66,30 @@ public class ChestMenu extends AbstractContainerMenu {
         return new ChestMenu(MenuType.GENERIC_9x6, containerId, inventory, 6);
     }
 
+    // Purpur start - Barrels and enderchests 6 rows
+    public static ChestMenu oneRow(int syncId, Inventory playerInventory, Container inventory) {
+        return new ChestMenu(MenuType.GENERIC_9x1, syncId, playerInventory, inventory, 1);
+    }
+
+    public static ChestMenu twoRows(int syncId, Inventory playerInventory, Container inventory) {
+        return new ChestMenu(MenuType.GENERIC_9x2, syncId, playerInventory, inventory, 2);
+    }
+
+    // Purpur end - Barrels and enderchests 6 rows
     public static ChestMenu threeRows(int containerId, Inventory inventory, Container container) {
         return new ChestMenu(MenuType.GENERIC_9x3, containerId, inventory, container, 3);
     }
 
+    // Purpur start - Barrels and enderchests 6 rows
+    public static ChestMenu fourRows(int syncId, Inventory playerInventory, Container inventory) {
+        return new ChestMenu(MenuType.GENERIC_9x4, syncId, playerInventory, inventory, 4);
+    }
+
+    public static ChestMenu fiveRows(int syncId, Inventory playerInventory, Container inventory) {
+        return new ChestMenu(MenuType.GENERIC_9x5, syncId, playerInventory, inventory, 5);
+    }
+
+    // Purpur end - Barrels and enderchests 6 rows
     public static ChestMenu sixRows(int containerId, Inventory inventory, Container container) {
         return new ChestMenu(MenuType.GENERIC_9x6, containerId, inventory, container, 6);
     }
diff --git a/net/minecraft/world/inventory/PlayerEnderChestContainer.java b/net/minecraft/world/inventory/PlayerEnderChestContainer.java
index 401b39d89cc485f6c237a8b7035690ece74eecfd..21c2472545d2fb0405e6582e6433986618e63569 100644
--- a/net/minecraft/world/inventory/PlayerEnderChestContainer.java
+++ b/net/minecraft/world/inventory/PlayerEnderChestContainer.java
@@ -26,11 +26,18 @@ public class PlayerEnderChestContainer extends SimpleContainer {
     }
 
     public PlayerEnderChestContainer(Player owner) {
-        super(27);
+        super(io.canvasmc.canvas.Config.INSTANCE.containers.enderChestSixRows ? 54 : 27); // Purpur - Barrels and enderchests 6 rows
         this.owner = owner;
         // CraftBukkit end
     }
 
+    // Purpur start - Barrels and enderchests 6 rows
+    @Override
+    public int getContainerSize() {
+        return owner.sixRowEnderchestSlotCount < 0 ? super.getContainerSize() : owner.sixRowEnderchestSlotCount;
+    }
+
+    // Purpur end - Barrels and enderchests 6 rows
     public void setActiveChest(EnderChestBlockEntity activeChest) {
         this.activeChest = activeChest;
     }
diff --git a/net/minecraft/world/level/block/EnderChestBlock.java b/net/minecraft/world/level/block/EnderChestBlock.java
index b0442a21f3bc41495a2cf34a407ed24bd30b93dc..610359a3d266eed124e682b52ac5cc7a61e355c6 100644
--- a/net/minecraft/world/level/block/EnderChestBlock.java
+++ b/net/minecraft/world/level/block/EnderChestBlock.java
@@ -85,8 +85,8 @@ public class EnderChestBlock extends AbstractChestBlock<EnderChestBlockEntity> i
                 enderChestInventory.setActiveChest(enderChestBlockEntity); // Needs to happen before ChestMenu.threeRows as it is required for opening animations
                 if (level instanceof ServerLevel serverLevel && player.openMenu(
                     new SimpleMenuProvider(
-                        (containerId, playerInventory, player1) -> ChestMenu.threeRows(containerId, playerInventory, enderChestInventory), CONTAINER_TITLE
-                    )
+                        (containerId, playerInventory, player1) -> io.canvasmc.canvas.Config.INSTANCE.containers.enderChestSixRows ? getEnderChestSixRows(containerId, playerInventory, player, enderChestInventory) : ChestMenu.threeRows(containerId, playerInventory, enderChestInventory), CONTAINER_TITLE
+                    ) // Purpur - Barrels and enderchests 6 rows
                 ).isPresent()) {
                 // Paper end - Fix InventoryOpenEvent cancellation - moved up;
                     player.awardStat(Stats.OPEN_ENDERCHEST);
@@ -100,6 +100,35 @@ public class EnderChestBlock extends AbstractChestBlock<EnderChestBlockEntity> i
         }
     }
 
+    // Purpur start - Barrels and enderchests 6 rows
+    private ChestMenu getEnderChestSixRows(int syncId, net.minecraft.world.entity.player.Inventory inventory, Player player, PlayerEnderChestContainer playerEnderChestContainer) {
+        if (io.canvasmc.canvas.Config.INSTANCE.containers.enderChestPermissionRows) {
+            org.bukkit.craftbukkit.entity.CraftHumanEntity bukkitPlayer = player.getBukkitEntity();
+            if (bukkitPlayer.hasPermission("purpur.enderchest.rows.six")) {
+                player.sixRowEnderchestSlotCount = 54;
+                return ChestMenu.sixRows(syncId, inventory, playerEnderChestContainer);
+            } else if (bukkitPlayer.hasPermission("purpur.enderchest.rows.five")) {
+                player.sixRowEnderchestSlotCount = 45;
+                return ChestMenu.fiveRows(syncId, inventory, playerEnderChestContainer);
+            } else if (bukkitPlayer.hasPermission("purpur.enderchest.rows.four")) {
+                player.sixRowEnderchestSlotCount = 36;
+                return ChestMenu.fourRows(syncId, inventory, playerEnderChestContainer);
+            } else if (bukkitPlayer.hasPermission("purpur.enderchest.rows.three")) {
+                player.sixRowEnderchestSlotCount = 27;
+                return ChestMenu.threeRows(syncId, inventory, playerEnderChestContainer);
+            } else if (bukkitPlayer.hasPermission("purpur.enderchest.rows.two")) {
+                player.sixRowEnderchestSlotCount = 18;
+                return ChestMenu.twoRows(syncId, inventory, playerEnderChestContainer);
+            } else if (bukkitPlayer.hasPermission("purpur.enderchest.rows.one")) {
+                player.sixRowEnderchestSlotCount = 9;
+                return ChestMenu.oneRow(syncId, inventory, playerEnderChestContainer);
+            }
+        }
+        player.sixRowEnderchestSlotCount = -1;
+        return ChestMenu.sixRows(syncId, inventory, playerEnderChestContainer);
+    }
+
+    // Purpur end - Barrels and enderchests 6 rows
     @Override
     public BlockEntity newBlockEntity(BlockPos worldPosition, BlockState blockState) {
         return new EnderChestBlockEntity(worldPosition, blockState);
diff --git a/net/minecraft/world/level/block/entity/BarrelBlockEntity.java b/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
index 434f67e6dc3f2644ac79edebcf089c03249445a6..c3f8b28fbfae2c3c5e382e2e18faf2da35d36d8e 100644
--- a/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
@@ -59,7 +59,17 @@ public class BarrelBlockEntity extends RandomizableContainerBlockEntity {
     }
     // CraftBukkit end
     private static final Component DEFAULT_NAME = Component.translatable("container.barrel");
-    private NonNullList<ItemStack> items = NonNullList.withSize(27, ItemStack.EMPTY);
+    // Purpur start - Barrels and enderchests 6 rows
+    private NonNullList<ItemStack> items = NonNullList.withSize(switch (io.canvasmc.canvas.Config.INSTANCE.containers.barrelRows) {
+        case 6 -> 54;
+        case 5 -> 45;
+        case 4 -> 36;
+        case 2 -> 18;
+        case 1 -> 9;
+        default -> 27;
+    }, ItemStack.EMPTY);
+
+    // Purpur end - Barrels and enderchests 6 rows
     public final ContainerOpenersCounter openersCounter = new ContainerOpenersCounter() {
         @Override
         protected void onOpen(Level level, BlockPos pos, BlockState blockState) {
@@ -111,7 +121,16 @@ public class BarrelBlockEntity extends RandomizableContainerBlockEntity {
 
     @Override
     public int getContainerSize() {
-        return 27;
+        // Purpur start - Barrels and enderchests 6 rows
+        return switch (io.canvasmc.canvas.Config.INSTANCE.containers.barrelRows) {
+            case 6 -> 54;
+            case 5 -> 45;
+            case 4 -> 36;
+            case 2 -> 18;
+            case 1 -> 9;
+            default -> 27;
+        };
+        // Purpur end - Barrels and enderchests 6 rows
     }
 
     @Override
@@ -131,7 +150,16 @@ public class BarrelBlockEntity extends RandomizableContainerBlockEntity {
 
     @Override
     protected AbstractContainerMenu createMenu(int containerId, Inventory inventory) {
-        return ChestMenu.threeRows(containerId, inventory, this);
+        // Purpur start - Barrels and enderchests 6 rows
+        return switch (io.canvasmc.canvas.Config.INSTANCE.containers.barrelRows) {
+            case 6 -> ChestMenu.sixRows(containerId, inventory, this);
+            case 5 -> ChestMenu.fiveRows(containerId, inventory, this);
+            case 4 -> ChestMenu.fourRows(containerId, inventory, this);
+            case 2 -> ChestMenu.twoRows(containerId, inventory, this);
+            case 1 -> ChestMenu.oneRow(containerId, inventory, this);
+            default -> ChestMenu.threeRows(containerId, inventory, this);
+        };
+        // Purpur end - Barrels and enderchests 6 rows
     }
 
     @Override
